### PR TITLE
Update SkillDialog to handle all activity types (e.g. set DeliveryMode for Invokes)

### PR DESCRIPTION
**Pull request by stevengum**
Fixes https://github.com/microsoft/botbuilder-js/issues/2139

## Specific Changes
  - Add `StatusCodes.MULTIPLE_CHOICES = 300`
  - Add protected method `SkillDialog.onValidateActivity()`
  - `SkillDialog` now forwards all `ActivityTypes` to the Skill.
  - Add new C# SkillDialog tests and refactor old tests
  - Add missing test for TurnContext when deliveryMode === 'expectReplies'
  - Fix error message in `BotFrameworkAdapter.createConnectorClientWithIdentity()`
  - Add tests for `BotFrameworkAdapter.createConnectorClientWithIdentity()`
  - Add test for `BotFrameworkAdapter.processActivity()` when deliveryMode === 'expectReplies', type is `Invoke` and reply activities have type `invokeResponse`

### CODE REVIEW

File: libraries/botbuilder-dialogs/src/skillDialog.ts [hunk 1]
```diff
@@ -162,35 +177,18 @@ export class SkillDialog extends Dialog {
...
-        // Create a conversationId to interact with the skill and send the activity
-        let skillConversationId: string;
-        try {
-            skillConversationId = await this.dialogOptions.conversationIdFactory.createSkillConversationIdWithOptions(conversationIdFactoryOptions);
-        } catch (err) {
-            if (err.message !== 'Not Implemented') throw err;
-            // If the SkillConversationIdFactoryBase implementation doesn't support createSkillConversationIdWithOptions(),
-            // use createSkillConversationId() instead.
-            skillConversationId = await this.dialogOptions.conversationIdFactory.createSkillConversationId(TurnContext.getConversationReference(activity) as ConversationReference);
+        if (activity.type === ActivityTypes.Invoke) {
```
Comment (johnataylor) [inline libraries/botbuilder-dialogs/src/skillDialog.ts:184]: having to switch expectReplies on for an invoke seems a little odd.

File: libraries/botbuilder/src/botFrameworkAdapter.ts [hunk 2]
```diff
@@ -843,8 +843,18 @@ export class BotFrameworkAdapter extends BotAdapter implements ConnectorClientBu
             context.turnState.set(BotCallbackHandlerKey, logic);
             await this.runMiddleware(context, logic);
 
-            // Retrieve cached invoke response.
-            if (request.type === ActivityTypes.Invoke) {
```
Comment (johnataylor) [inline libraries/botbuilder/src/botFrameworkAdapter.ts:847]: and this change presumably for consistency with the other place this condition occurs...

Comment (stevengum) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:847]: Yes, these changes are ported from C# with exactly the same spirit (and tests) in mind. That said, I will add some more complex unit tests tomorrow to cover the interactions between the BFAdapter and ActivityHandler, and the two with SkillDialog.

File: libraries/botbuilder-dialogs/src/skillDialog.ts (hunk 1 above)
Comment (gabog) [reply libraries/botbuilder-dialogs/src/skillDialog.ts:184]: if the dialog is interacting with a skill that only works with invokes, this makes sure that EoC replies are buffered with the invokeResponse and the dialog ends when an EoC comes back.

**Pull request merged**